modules/core/shared/src/main/scala-2/weaver/ExpectMacro.scala

```diff
@@ -0,0 +1,162 @@
+                        cluesValDef,
+                        List(transformedValue),
+                        sourceLoc,
+                        q"Some($message)")
+  }
+
+  /**
+   * Constructs [[Expectations]] from a boolean value.
+   *
+   * A macro is needed to support clues. The value expression may contain calls
+   * to [[ClueHelpers.clue]], which generate clues for values under test.
+   *
+   * This macro constructs a local collection of [[Clues]] and adds the
+   * generated clues to it. Calls to [[ClueHelpers.clue]] are rewritten to calls
+   * to [[Clues.addClue]].
+   *
+   * After the value is evaluated, the [[Clues]] collection is used to contruct
+   * [[Expectations]].
+   */
+  def applyImpl(c: blackbox.Context)(value: c.Tree): c.Tree = {
+
+    import c.universe._
+    val sourceLoc = new weaver.macros.Macros(c).fromContext.asInstanceOf[c.Tree]
+    val (cluesName, cluesValDef) = makeClues(c)
+    val clueMethodSymbol         = getClueMethodSymbol(c)
+
+    val transformedValue =
+      replaceClueMethodCalls(c)(clueMethodSymbol, cluesName, value)
+    makeExpectations(c)(cluesName,
+                        cluesValDef,
+                        List(transformedValue),
+                        sourceLoc,
+                        q"None")
+  }
+
+  /** Constructs [[Expectations]] from the local [[Clues]] collection. */
+  private def makeExpectations(c: blackbox.Context)(
+      cluesName: c.TermName,
+      cluesValDef: c.Tree,
+      values: List[c.Tree],
+      sourceLoc: c.Tree,
+      message: c.Tree): c.Tree = {
+    import c.universe._
+    val block =
+      q"$cluesValDef; _root_.weaver.internals.Clues.toExpectations($sourceLoc, $message, $cluesName, ..$values)"
+    val untyped = c.untypecheck(block)
+    val retyped = c.typecheck(untyped, pt = c.typeOf[Expectations])
+    retyped
+
+  }
+
+  /** Get the [[ClueHelpers.clue]] symbol. */
+  private def getClueMethodSymbol(c: blackbox.Context): c.Symbol = {
+    import c.universe._
+    symbolOf[ClueHelpers].info.member(TermName("clue"))
+  }
+
+  /** Construct a [[Clues]] collection local to the `expect` call. */
+  private def makeClues(c: blackbox.Context): (c.TermName, c.Tree) = {
+    import c.universe._
+    val cluesName = TermName(c.freshName("clues$"))
+    val cluesValDef =
+      q"val $cluesName: _root_.weaver.internals.Clues = new _root_.weaver.internals.Clues()"
+    (cluesName, cluesValDef)
+  }
+
+  /**
+   * Replaces all calls to [[ClueHelpers.clue]] with calls to [[Clues.addClue]].
+   */
+  private def replaceClueMethodCalls(c: blackbox.Context)(
+      clueMethodSymbol: c.Symbol,
+      cluesName: c.TermName,
+      value: c.Tree): c.Tree = {
+
+    import c.universe._
+    object transformer extends Transformer {
+
+      override def transform(input: Tree): Tree = input match {
+        case c.universe.Apply(fun, List(clueValue))
+            if fun.symbol == clueMethodSymbol =>
+          val transformedClueValue = super.transform(clueValue)
```

Comment:
I'm curious about this transformation here. Can you explain ?

Reply:
Of course!

For context, the user writes `clue(x)` where `clue` refers to `ClueHelpers.clue`.   There's an implicit conversion from `x` to a `Clue`. Explicitly, this is written as `clue(Clue.generateClue(x))`  . 

The transformation replaces `ClueHelpers.clue` with `addClue` on a clues collection.  In other words, it replaces:
`clue(Clue.generateClue(x))` with `clues.addClue(Clue.generateClue(x))`.

The clues collection is generated as part of the overall `expect` macro. For example:

```scala
expect(clue(List(x, clue(y))))
```
becomes

```scala
expect {
  val clues = new Clues()
  clues.addClue(List(x, clues.addClue(y)))
}
```